File: wi-seong-cheol/BOJ/0x08 Stack2/9012.cpp
//
//  9012.cpp
//  main
//
//  Created by wi_seong on 2022/11/10.
//

/*
 [Input]
 6
 (())())
 (((()())()
 (()())((()))
 ((()()(()))(((())))()
 ()()()()(()()())()
 (()((())()(
 [Output]
 NO
 NO
 YES
 NO
 YES
 NO
 */
// 시간복잡도: O(n)
// 최악시간: 50 * T
// 난이도: Silver 4
// Timer: 3m
// Url: https://www.acmicpc.net/problem/9012

#include <iostream>
#include <string>
#include <stack>

using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    int n; cin >> n;
    while(n--) {
        stack<char> S;
        string s; cin >> s;
        for(auto ch: s) {
            if(ch == '(') S.push(ch);
            else {
                if(!S.empty() && S.top() == '(') S.pop();
                else {
                    S.push(ch);
                    break;
                }
            }
        }
        if(S.empty()) cout << "YES\n";
        else cout << "NO\n";
    }
    return 0;
}

File: wlwl1011/BOJ/heybob/4회/9012.py
import sys
input = lambda : sys.stdin.readline().rstrip('\r\n')

for i in range(int(input())):
    s = list(input())
    stack = []
    for i in range(len(s)):
        if s[i] == '(':
            stack.append('(')
        else:
            if stack and stack[-1] == '(':
                stack.pop() 
            else:
                stack.append(s[i])     
    if stack:
        print('NO')
    else:
        print('YES')                  

